### PR TITLE
External TURN with REST auth

### PR DESCRIPTION
**Pull request by athoune**
[Coturn](https://github.com/coturn/coturn) implements http://tools.ietf.org/html/draft-uberti-behave-turn-rest-00 lets use it

### CODE REVIEW

File: turn/turn_rest.go
Comment (jmattheis) [inline turn/turn_rest.go:13]: Hey, thanks for your contribution!

Will the implementation be done in this PR?

Comment (athoune) [reply turn/turn_rest.go]: I'm struggling with a clean interface, usable with embeded turn service and coturn : turn/turn.go

I'm implementing the authentication with hmac : turn/turn_rest.go

File: turn/turn_rest.go
Comment (maximbaz) [inline turn/turn_rest.go:39]: This needs to be base64-encoded, i.e. 

```go
account.Credential = base64.StdEncoding.EncodeToString(mac.Sum(nil))
``` 

(and import `"encoding/base64"`)

**Review by maximbaz — COMMENTED** [review]
Thanks for your work @athoune! I know you are still working on this but I wanted to help you out with some early feedback, I tested and can confirm that with my suggested changes I got it working, both `screego` and `coturn` running in docker, and with `STUN` completely disabled in `coturn` I can confirm that screen sharing is working as expected via `TURN` 👍 

Looking forward to see this completed, would also be happy to help with further testing and code review once you think it's ready for some feedback! 🙂

File: config/config.go
Comment (jmattheis) [inline config/config.go:47]: Just a little hint for testing, there is a hidden url parameter [`forceTurn=true`](https://github.com/screego/server/blob/ab187c14f3934d0e9fd3ddc2ff54ced7c93a1042/ui/src/useRoom.ts#L36) which forces the usage of turn, even if Stun would be enough. Example: https://app.screego.net/?room=sound-lavender-badger&forceTurn=true

File: config/config.go
Comment (jmattheis) [reply config/config.go:50]: Also, instead of a boolean flag, it probably would be cool, if you could specify the address of the external turn server, so that it can be hosted on a separate server.

File: turn/server.go
Comment (jmattheis) [inline turn/server.go:116]: This seems a little weird, 2 users with the same username but different credentials, is this valid?

File: turn/turn_rest.go
Comment (jmattheis) [inline turn/turn_rest.go:55]: Secret and port shoudn't be hardcoded.

File: config/config.go
Comment (athoune) [reply config/config.go]: See 799f05e4e6a8b45e6b1aab1faacbd4d743206fc5

File: turn/turn_rest.go
Comment (athoune) [reply turn/turn_rest.go]: See 1ebe9ab6cd972fa4d4c45091d35d25ab86185990

File: turn/server.go
Comment (athoune) [reply turn/server.go:116]: Oups, it's a mistake.

File: turn/turn_rest.go
Comment (athoune) [reply turn/turn_rest.go]: Oups, I left a `FIXME`.

See 111041358df79a25c7812dc3e6b0283844dbd1ac

File: config/config.go
Comment (maximbaz) [reply config/config.go]: @jmattheis it's essentially `ExternalIP` config that points to the location of external TURN server, right? it's the only place where it is used, in `ws/room.go` in `addresses()`

In my setup coturn and screego are not in the same network (but both have the same public IP), and this works 🙂

File: config/config.go
Comment (maximbaz) [inline config/config.go:52]: I would suggest avoiding using "coturn" in config name, how about `TurnExternalSecret`?

To be honest I am not sure if we even need to expose TTL param, we could try to be pragmatic and leave it hardcoded to 24h as suggested in the [docs](http://tools.ietf.org/html/draft-uberti-behave-turn-rest-00) and wait until someone actually has a need to configure this? 🤔

File: turn/turn_rest.go
Comment (maximbaz) [inline turn/turn_rest.go:69]: It's a little unfortunate that we need to specify addresses pointing to TURN server (IPv4 and IPv6) in `ExternalIP` parameter (which is used in `ws/room.go` in `addresses()`, and port to the TURN server in `TurnAddress`, where the IP address of `TurnAddress` is practically useless...

Not sure what you guys think about it, is it just a matter of properly documenting this, or can you think of any other way to configure this?

Maybe `TurnAddress` can be parsed in a similar way as `ExternalIP` (i.e. it can take both IPv4 and IPv6 values), and then add `IpAddressV4` and `IpAddressV6` to `TurnServer` struct, so that in `ws/room.go::addresses()` we would use like below?

```go
fmt.Sprintf("%s:%s:%d?transport=tcp", prefix, r.turnServer.IpAddressV4(), r.turnServer.Port()))
```

File: config/config.go
Comment (athoune) [reply config/config.go:52]: coturn seems to be the only TURN server using the hmac trick for generating password.

* [ ] Hardcode TTL to 24h

Comment (maximbaz) [reply config/config.go:51]: Since coturn is the leading implementation I think they are the only ones who do many things 😄 I just thought since it's a proposed standard for REST authentication, documented not on coturn's website but on ietf.org, others _might_ implement this as well at some point 🤷‍♂️

Comment (athoune) [reply config/config.go]: Fixed in 7ba0af398f0c60351c1df022c38689900a05e59b

Comment (maximbaz) [reply config/config.go]: Awesome, thanks! Super excited to see this PR becoming more and more finished 👍 

Just for my understanding, I'm very new to TURN, I see you use "signed secret" term, does that mean that the shared secret is already signed by coturn, or that we sign something else using this secret? It's not a comment, I'm just curious how this works internally, coturn calls this value "auth secret" and the doc you mentioned in the first message calls it "shared secret key", so I would assume that this secret is not signed, but is used for signing?

**Comment by athoune** [comment]
I just rebase master.

File: turn/server.go
Comment (athoune) [reply turn/server.go]: `client` and `host` use their own Id for username.

The `AcceptAccounts` use now a variadic argument and accept in a loop.

Is it correct for you, @jmattheis ?

File: turn/turn_rest.go
Comment (athoune) [reply turn/turn_rest.go:67]: OK, it take time, I'm bit tired, but I understand your point.

`config.Config.ExternalIP` is only used for the turn server (without port), and `config.Config.TurnAddress` is a full address but only use the port part.

`config.Config.ServerAddress` is used for the web service (ip+port).

File: turn/server.go
Comment (jmattheis) [reply turn/server.go]: Yeah, maybe we could remove the AcceptAccounts / Revoke methods & turn.account and just pass the username & password like it is done in the allow method (and call allow 2 times on session creation)? I feel like creating a struct (Account) that is only used in one method is overkill.

File: turn/turn_rest.go
Comment (jmattheis) [inline turn/turn_rest.go:57]: AcceptAccounts

File: ws/room.go
Comment (jmattheis) [inline ws/room.go:75]: Do I understand this correctly: If turn external is enabled, then external ip should point to the ip of the turn server? As there may not be on the same server?

Comment (maximbaz) [reply ws/room.go:75]: This is correct, the `ExternalIP` address is basically used not to point to where screego itself is, but where the STUN/TURN server is - previously as STUN/TURN was always part of screego the IP address would be the address of where screego is hosted, but with this PR it becomes just a special case of the full range of possibilities 🙂

Maybe we should keep `SCREEGO_SERVER_ADDRESS=0.0.0.0:5050` and `SCREEGO_TURN_ADDRESS=0.0.0.0:3478` dedicated to screego itself (where it should start listening), but `SCREEGO_EXTERNAL_IP` should really become "list of external IP **and port** of the STUN/TURN server", and so this `addresses` function would only use this parameter alone?

@athoune what do you think?

Comment (athoune) [reply ws/room.go:75]: handling multiple external turn server with their secret can be complex. Does Screego need a cluster of turn servers?

Comment (maximbaz) [reply ws/room.go:75]: ah no I meant a single TURN server (with single secret) but with possibly multiple combinations of IP and ports to reach it, for example coturn can listen on IPv4 and IPv6, for some scenarios it can actually have two IPv4, it can also run on multiple ports just because some ports can be blocked by firewall, so it would be nice to be able to specify all these combinations - but it's still a single TURN server 🙂

Comment (athoune) [reply ws/room.go:75]: `SCREEGO_EXTERNAL_IP` with multiple values should be implemented in another PR, IMHO.

Comment (athoune) [reply ws/room.go:75]: What is missing for approving this PR?

File: ws/room.go
Comment (jmattheis) [inline ws/room.go:62]: This probably shouldn't panic

File: turn/turn.go
Comment (jmattheis) [inline turn/turn.go:9]: See https://github.com/screego/server/pull/63#discussion_r559190169

**Review by jmattheis — CHANGES_REQUESTED** [review]
Sorry, kinda overlooked this PR. 2 remarks left.

**Comment by maximbaz** [comment]
Hello guys, a friendly ping 🙂 It seems we only have a couple of minor things left, shall we try to get this merged?

**Comment by maximbaz** [comment]
Hello @athoune, if you are busy may I offer to finish this PR? Seems like it's just a coupe of remarks left. @jmattheis since I dont have write access to this branch, would that be okay if I just send a `.patch` file that you could apply on top of this branch and merge this all?

**Comment by jmattheis** [comment]
@maximbaz yes, you can send me a patch file.

**Pull request closed without merging**
